### ISSUE TITLE
sort out reporting of sample sizes for Dirichlet-Multinomial likelihood

### ISSUE THREAD

**Issue by taylori** [r4ss/r4ss#129]
D-M option in 3.30 needs update to r4ss functions 
This may need to all happen within SS itself as r4ss mostly depends on output of sample sizes in CompReport.sso.
@James-Thorson has provided equations by email,

**Comment by taylori** [comment]
Dirichlet-Multinomial not yet fully documented in SS user manual and seems less likely to get used in this assessment cycle, so removing from "2017 assessment cycle" milestone. If @James-Thorson has time to contribute, the next steps are probably something like the following:

- Document the additional SS inputs required to use this option.
- Modify calculation of effective sample size in SS to match Equation 11 from Thorson et al. (2016): 
`N_eff = 1 / (1+Theta) + N * Theta / (1+Theta)` (where `Theta` is the estimated DM parameter and `N` is the input sample size.  SS estimates the log of the DM parameter.)
- Make whatever changes are needed to SS to adapt to this option. For instance, would it be better to no longer report suggested Francis or McAllister-Ianelli weights for comparison since the likelihood is no longer multinomial.

Reference:
Thorson, J. T., Johnson, K. F., Methot, R. D., & Taylor, I. G. (2016). Model-based estimates of effective sample size in stock assessment models using the Dirichlet-multinomial distribution. Fisheries Research. https://doi.org/10.1016/j.fishres.2016.06.005

**Comment by James-Thorson** [comment]
Ian -- can you point me to the lines of code in r4ss where I would add code to calculate N_eff?  This task is a bit daunting for me, as I've never really done a deep-dive on r4ss code.

jim

**Comment by taylori** [comment]
The following change were made in commit 8c28eef to the [development](https://github.com/r4ss/r4ss/tree/development) branch, which can be installed via 
`devtools::install_github('r4ss/r4ss', ref='development')`

After discussion with @James-Thorson and @kellijohnson, I've just revised the output to report the Dirichlet-Multinomial parameters (estimated on a log scale), along with the associated Theta values and the ratio of Theta / (1 + Theta) which is an approximation of the sample size multiplier (and a replacement for the iterative tuning associated with the McAllister-Ianelli or Francis data-weighting methods for composition data).
```
$Dirichlet_Multinomial_pars
                    Value Phase Min Max      Theta Theta/(1+Theta)
ln(EffN_mult)_1 -0.221926     2  -5  20  0.8009746       0.4447451
ln(EffN_mult)_2  2.351210     2  -5  20 10.4982649       0.9130304
```

The sample sizes in the age comp plots have also be changed to reflect the new likelihood and the caption much improved to reflect what's represented (example below):

WARNING: changes haven't yet been implemented for length comps and may not yet work right for conditional age-at-length comps.

![image](https://user-images.githubusercontent.com/4992918/35069923-7a005722-fb90-11e7-8559-d4bdf96234e6.png)

**Comment by aaronmberger** [comment]
SSgetoutput breaks when internally calling SS_output when using the Dirichlet-Multinomial weighting approach as it returns an undefined variable (agebase) error:
'Error in nrow(agedbase) : object 'agedbase' not found'

Stumbled upon this when loading various runs for creating likelihood profiles as in:
profilemodels <- SSgetoutput(dirvec=file.path(path.in, folder.name), keyvec=1:length(R0vec),getcovar=FALSE,getcomp = FALSE)

If you set getcomp=TRUE, everything loads and operates as anticipated, so likely a variable accounting issue when running models that use the Dirichlet-Multinomial weighting approach.

**Comment by taylori** [comment]
Thanks for reporting this @aaronmberger. The issue was for any call to SS_output where the CompReport.sso either wasn't available or was skipped to run faster. It should be fixed with commit 02057f5 to the development branch, available as `devtools::install_github('r4ss/r4ss', ref='development')`. 

I also added some additional uses of `verbose` in commit 4cd2cfb so that SSgetoutput suppresses some of the messages related to the Dirichlet-Multinomial as it does for other elements of reading the output.

**Comment by taylori** [comment]
This should now be fully functional in the `development` branch of r4ss. Closing issue. We can always opena  new issue of there is any issue with models that use the Dirichlet-Multinomial likelihood in the future.
